code/crates/core-driver/src/driver.rs

```diff
@@ -395,15 +438,39 @@ where
             return Ok(None);
         };
 
-        let round_input = self.multiplex_vote_threshold(output, vote_round);
+        if let VKOutput::PolkaValue(val) = &output {
+            self.store_polka_certificate(vote_round, val);
+        }
 
+        let round_input = self.multiplex_vote_threshold(output, vote_round);
         if round_input == RoundInput::NoInput {
             return Ok(None);
         }
 
         self.apply_input(vote_round, round_input)
     }
 
+    fn store_polka_certificate(&mut self, vote_round: Round, value_id: &ValueId<Ctx>) {
+        let Some(per_round) = self.vote_keeper.per_round(vote_round) else {
+            return;
+        };
+
+        self.polka_certificates.push(PolkaCertificate {
+            height: self.height(),
+            round: vote_round,
+            value_id: value_id.clone(),
+            votes: per_round
```

Comment:
TDOO: Like in the [Quint spec](https://github.com/informalsystems/malachite/pull/364/files#diff-ef09dc528370554b2e2e64658d529bd0666761d785a190f2fe6bd0105fa719ccR545), we could double-check that there is enough voting power in this polka, but maybe that's overkill?